lib/siftsciex/event/login.ex

```diff
@@ -0,0 +1,171 @@
+defmodule Siftsciex.Event.Login do
+  @moduledoc """
+  Functions for managing the login event on Sift Science.
+  """
+  import Siftsciex
+
+  alias Siftsciex.Event.Payload
+  alias Siftsciex.Event.Payload.{App, Browser}
+
+  defstruct "$type": "$login",
+            "$api_key": api_key(),
+            "$user_id": :empty,
+            "$session_id": :empty,
+            "$login_status": :empty,
+            "$user_email": :empty,
+            "$ip": :empty,
+            "$browser": :empty,
+            "$app": :empty,
+            "$failure_reason": :empty,
+            "$username": :empty,
+            "$social_sign_on_type": :empty,
+            "$account_types": :empty,
+            "$brand_name": :empty,
+            "$site_country": :empty,
+            "$site_domain": :empty
+
+  @type t :: %__MODULE__{
+          "$type": String.t(),
+          "$api_key": String.t(),
+          "$user_id": String.t(),
+          "$session_id": Payload.payload_string(),
+          "$login_status": String.t(),
+          "$user_email": Payload.payload_string(),
+          "$ip": Payload.payload_string(),
+          "$browser": :empty,
+          "$app": :empty,
+          "$failure_reason": :empty,
+          "$username": :empty,
+          "$social_sign_on_type": :empty,
+          "$account_types": :empty,
+          "$brand_name": :empty,
+          "$site_country": :empty,
+          "$site_domain": :empty
```

Comment:
Why are these all `:empty`? My elixir knowledge is not the best, but does this not mean they won't be supported and if someone tries to use them to send these fields to Sift it won't allow them to?

Reply:
I don't know why the original author did it this way, but any entry with `:empty` as its value is removed before the final json is generated. Also, according to the API spec all these fields are optional.

Reply:
I'm confused, where are you seeing this from the original author? I'm not seeing any other code in here that has a bunch of `:empty` values in the type definitions

Reply:
Look at any `Siftsciex.Event.Payload.*` module and you'll see. Examples [here](https://github.com/msramos/siftsciex/blob/e2861aca2e20a03bd081c98c3a377f722c033d3a/lib/siftsciex/event/payload/promotion.ex#L12-L12), [here](https://github.com/msramos/siftsciex/blob/e2861aca2e20a03bd081c98c3a377f722c033d3a/lib/siftsciex/event/payload/payment_method.ex#L12-L12) and [here](https://github.com/msramos/siftsciex/blob/e2861aca2e20a03bd081c98c3a377f722c033d3a/lib/siftsciex/event/payload/item.ex#L9-L9).

You can see the `purge_empty` function [here](https://github.com/msramos/siftsciex/blob/e2861aca2e20a03bd081c98c3a377f722c033d3a/lib/siftsciex/event.ex#L269-L285)

Reply:
Oh I got it. You mean in the _typespec_ itself.

Yeah, you're right! I'll fix this.

But FYI specs are just docs, they will not interfere with the runtime and are only used by static analysis.